quaddtype/numpy_quaddtype/src/dtype.c

```diff
@@ -0,0 +1,261 @@
+#include <Python.h>
+#include <sleef.h>
+#include <sleefquad.h>
+
+#define PY_ARRAY_UNIQUE_SYMBOL QuadPrecType_ARRAY_API
+#define PY_UFUNC_UNIQUE_SYMBOL QuadPrecType_UFUNC_API
+#define NPY_NO_DEPRECATED_API NPY_2_0_API_VERSION
+#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
+#define NO_IMPORT_ARRAY
+#define NO_IMPORT_UFUNC
+#include "numpy/arrayobject.h"
+#include "numpy/ndarraytypes.h"
+#include "numpy/dtype_api.h"
+
+#include "scalar.h"
+#include "casts.h"
+#include "dtype.h"
+#include "dragon4.h"
+
+static inline int
+quad_load(void *x, char *data_ptr, QuadBackendType backend)
+{
+    if (data_ptr == NULL || x == NULL) {
+        return -1;
+    }
+    if (backend == BACKEND_SLEEF) {
+        *(Sleef_quad *)x = *(Sleef_quad *)data_ptr;
+    }
+    else {
+        *(long double *)x = *(long double *)data_ptr;
+    }
+    return 0;
+}
+
+static inline int
+quad_store(char *data_ptr, void *x, QuadBackendType backend)
+{
+    if (data_ptr == NULL || x == NULL) {
+        return -1;
+    }
+    if (backend == BACKEND_SLEEF) {
+        *(Sleef_quad *)data_ptr = *(Sleef_quad *)x;
+    }
+    else {
+        *(long double *)data_ptr = *(long double *)x;
+    }
+    return 0;
+}
+
+QuadPrecDTypeObject *
+new_quaddtype_instance(QuadBackendType backend)
+{
+    QuadBackendType target_backend = backend;
+    if (backend != BACKEND_SLEEF && backend != BACKEND_LONGDOUBLE) {
+        PyErr_SetString(PyExc_TypeError, "Backend must be sleef or longdouble");
+        return NULL;
+        // target_backend = BACKEND_SLEEF;
+    }
+
+    QuadPrecDTypeObject *new = (QuadPrecDTypeObject *)PyArrayDescr_Type.tp_new(
+            (PyTypeObject *)&QuadPrecDType, NULL, NULL);
+    if (new == NULL) {
+        return NULL;
+    }
+    new->base.elsize = (target_backend == BACKEND_SLEEF) ? sizeof(Sleef_quad) : sizeof(long double);
+    new->base.alignment =
+            (target_backend == BACKEND_SLEEF) ? _Alignof(Sleef_quad) : _Alignof(long double);
+    new->backend = target_backend;
+    return new;
+}
+
+static QuadPrecDTypeObject *
+ensure_canonical(QuadPrecDTypeObject *self)
+{
+    Py_INCREF(self);
+    return self;
+}
+
+static QuadPrecDTypeObject *
+common_instance(QuadPrecDTypeObject *dtype1, QuadPrecDTypeObject *dtype2)
+{
+    if (dtype1->backend != dtype2->backend) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Cannot find common instance for QuadPrecDTypes with different backends");
+        return NULL;
```

Comment:
This is fine but if anyone mentions it in the future, I think it would probably also work to return a SLEEF quad, since it's always safe to cast a long double to a quad.